src/memo/memo.ml

```diff
@@ -647,6 +666,55 @@ module Call_stack = struct
     Fiber.Var.set call_stack_var stack (fun () -> Implicit_output.forbid f)
 end
 
+module Error_handler : sig
+  val is_set : bool Fiber.t
+
+  val report_error : Exn_with_backtrace.t -> unit Fiber.t
+
+  val with_error_handler :
+    (Exn_with_backtrace.t -> unit Fiber.t) -> (unit -> 'a Fiber.t) -> 'a Fiber.t
+end = struct
+  type t = Exn_with_backtrace.t -> unit Fiber.t
+
+  let var : t Fiber.Var.t = Fiber.Var.create ()
+
+  let is_set = Fiber.map (Fiber.Var.get var) ~f:Option.is_some
+
+  let get_exn = Fiber.Var.get_exn var
+
+  let report_error error =
+    let open Fiber.O in
+    let* handler = get_exn in
+    let* stack = Call_stack.get_call_stack_without_state () in
+    let error =
+      Exn_with_backtrace.map error ~f:(fun exn ->
+          List.fold_left stack ~init:exn ~f:(fun exn stack_frame ->
+              Error.extend_stack exn ~stack_frame))
+    in
+    handler error
```

Comment:
I wonder if we should raise a code error here if the handler disobeys the contract and raises an exception.

Reply:
I just pushed code that does this, but, guess what, raising a code error amounts to raising an exception, so it won't prevent confusing errors. It will hopefully add a less confusing error to the confusing ones.